File: lang/typescript/data_bin_one_shots_test.ts
#!/usr/bin/env tsx
/**
 * Round-trip tests for the TypeScript data_bin one-shot wrappers
 * (Phase 7.28 V core; Phase 7.33 TypeScript binding).
 *
 * Loaders return UNFRAMED payload Buffers (matching toDataBin's
 * convention; callBinFn strips the [u32 LE size] frame). Dumpers
 * expect FRAMED Buffer input (matching fromDataBin's convention).
 * Tests use a `reframe()` helper to bridge.
 */
import {
  xmlToDataBin, jsonToDataBin, yamlToDataBin, tomlToDataBin, mdToDataBin,
  dataBinToXml, dataBinToJson, dataBinToYaml, dataBinToToml, dataBinToMd,
} from './cxlib/src/index';

function reframe(payload: Buffer): Buffer {
  const out = Buffer.alloc(4 + payload.length);
  out.writeUInt32LE(payload.length, 0);
  payload.copy(out, 4);
  return out;
}

let _passed = 0;
let _failed = 0;

function _run(name: string, fn: () => void): void {
  try {
    fn();
    _passed++;
  } catch (e: any) {
    _failed++;
    console.log(`  FAIL ${name}: ${e.message}`);
  }
}

// ── XML one-shot ─────────────────────────────────────────────────────────────

_run('xmlToDataBin returns CXDB payload', () => {
  const payload = xmlToDataBin('<server><host>localhost</host><port>8080</port></server>');
  if (payload.length < 4) throw new Error(`expected non-empty payload, got ${payload.length}`);
  // Magic check
  const magic = payload.subarray(0, 4).toString('ascii');
  if (magic !== 'CXDB') throw new Error(`expected CXDB magic, got ${magic}`);
});

_run('xml round-trip through data_bin', () => {
  const payload = xmlToDataBin('<server><host>localhost</host><port>8080</port></server>');
  const out = dataBinToXml(reframe(payload));
  if (!out.includes('server')) throw new Error(`expected server: ${out}`);
  if (!out.includes('localhost')) throw new Error(`expected localhost: ${out}`);
  if (!out.includes('8080')) throw new Error(`expected 8080: ${out}`);
});

// ── JSON one-shot ────────────────────────────────────────────────────────────

_run('json round-trip through data_bin', () => {
  const payload = jsonToDataBin('{"name": "alice", "id": 1}');
  const out = dataBinToJson(reframe(payload));
  if (!out.includes('alice')) throw new Error(`expected alice: ${out}`);
  if (!out.includes('1')) throw new Error(`expected 1: ${out}`);
});

// ── YAML one-shot ────────────────────────────────────────────────────────────

_run('yaml round-trip through data_bin', () => {
  const payload = yamlToDataBin('name: alice\nid: 1\n');
  const out = dataBinToYaml(reframe(payload));
  if (!out.includes('alice')) throw new Error(`expected alice: ${out}`);
});

// ── TOML one-shot ────────────────────────────────────────────────────────────

_run('toml round-trip through data_bin', () => {
  const payload = tomlToDataBin('name = "alice"\nid = 1\n');
  const out = dataBinToToml(reframe(payload));
  if (!out.includes('alice')) throw new Error(`expected alice: ${out}`);
});

// ── Markdown one-shot ────────────────────────────────────────────────────────

_run('md round-trip through data_bin', () => {
  const payload = mdToDataBin('# Title\n\nA paragraph.\n');
  const out = dataBinToMd(reframe(payload));
  if (!out.includes('Title')) throw new Error(`expected Title: ${out}`);
});

// ── Cross-format compositions ────────────────────────────────────────────────

_run('xml → data_bin → json', () => {
  const payload = xmlToDataBin('<user id="1" name="alice"/>');
  const out = dataBinToJson(reframe(payload));
  if (!out.includes('alice')) throw new Error(`expected alice: ${out}`);
  if (!out.includes('1')) throw new Error(`expected 1: ${out}`);
});

_run('json → data_bin → yaml', () => {
  const payload = jsonToDataBin('{"name": "alice", "active": true}');
  const out = dataBinToYaml(reframe(payload));
  if (!out.includes('alice')) throw new Error(`expected alice: ${out}`);
});

_run('toml → data_bin → xml', () => {
  const payload = tomlToDataBin('host = "localhost"\nport = 8080\n');
  const out = dataBinToXml(reframe(payload));
  if (!out.includes('localhost')) throw new Error(`expected localhost: ${out}`);
  if (!out.includes('8080')) throw new Error(`expected 8080: ${out}`);
});

// ── main ──────────────────────────────────────────────────────────────────────

const status = _failed === 0 ? 'OK' : 'FAILED';
console.log(`typescript/data_bin_one_shots_test.ts: ${_passed} passed, ${_failed} failed  [${status}]`);
process.exit(_failed === 0 ? 0 : 1);

File: lang/rust/cxlib/tests/data_bin_one_shots_test.rs
//! Round-trip tests for the Rust data_bin one-shot wrappers
//! (Phase 7.28 V core; Phase 7.32 Rust binding).
//!
//! Loaders return UNFRAMED payload bytes (matching `to_data_bin`'s
//! convention; the wrapper strips the [u32 LE size] frame). Dumpers
//! expect FRAMED input (matching `from_data_bin`'s convention). Tests
//! use a `reframe` helper to bridge.

use cxlib::data_bin::*;

fn reframe(payload: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(4 + payload.len());
    let size = payload.len() as u32;
    out.extend_from_slice(&size.to_le_bytes());
    out.extend_from_slice(payload);
    out
}

// ── XML one-shot ─────────────────────────────────────────────────────────────

#[test]
fn xml_to_data_bin_returns_cxcol_payload() {
    let payload = xml_to_data_bin("<server><host>localhost</host><port>8080</port></server>")
        .expect("xml_to_data_bin");
    assert!(payload.len() > 5, "expected non-empty payload, got {} bytes", payload.len());
    // Magic check — 5-byte "CXCol" per spec/core/data-bin.md §3.1 (v0.8.0).
    assert_eq!(&payload[..5], b"CXCol", "expected CXCol magic, got {:?}", &payload[..5]);
}

#[test]
fn xml_round_trip_through_data_bin() {
    let payload = xml_to_data_bin("<server><host>localhost</host><port>8080</port></server>")
        .expect("xml_to_data_bin");
    let out = data_bin_to_xml(&reframe(&payload)).expect("data_bin_to_xml");
    assert!(out.contains("server"), "expected server in xml output, got: {}", out);
    assert!(out.contains("localhost"), "expected localhost, got: {}", out);
    assert!(out.contains("8080"), "expected 8080, got: {}", out);
}

// ── JSON one-shot ────────────────────────────────────────────────────────────

#[test]
fn json_round_trip_through_data_bin() {
    let payload = json_to_data_bin(r#"{"name": "alice", "id": 1}"#).expect("json_to_data_bin");
    let out = data_bin_to_json(&reframe(&payload)).expect("data_bin_to_json");
    assert!(out.contains("alice"));
    assert!(out.contains("1"));
}

// ── YAML one-shot ────────────────────────────────────────────────────────────

#[test]
fn yaml_round_trip_through_data_bin() {
    let payload = yaml_to_data_bin("name: alice\nid: 1\n").expect("yaml_to_data_bin");
    let out = data_bin_to_yaml(&reframe(&payload)).expect("data_bin_to_yaml");
    assert!(out.contains("alice"));
}

// ── TOML one-shot ────────────────────────────────────────────────────────────

#[test]
fn toml_round_trip_through_data_bin() {
    let payload = toml_to_data_bin("name = \"alice\"\nid = 1\n").expect("toml_to_data_bin");
    let out = data_bin_to_toml(&reframe(&payload)).expect("data_bin_to_toml");
    assert!(out.contains("alice"));
}

// ── Cross-format compositions ────────────────────────────────────────────────

#[test]
fn xml_to_data_bin_to_json() {
    let payload = xml_to_data_bin(r#"<user id="1" name="alice"/>"#).expect("xml_to_data_bin");
    let out = data_bin_to_json(&reframe(&payload)).expect("data_bin_to_json");
    assert!(out.contains("alice"));
    assert!(out.contains("1"));
}

#[test]
fn json_to_data_bin_to_yaml() {
    let payload = json_to_data_bin(r#"{"name": "alice", "active": true}"#)
        .expect("json_to_data_bin");
    let out = data_bin_to_yaml(&reframe(&payload)).expect("data_bin_to_yaml");
    assert!(out.contains("alice"));
}

#[test]
fn toml_to_data_bin_to_xml() {
    let payload = toml_to_data_bin("host = \"localhost\"\nport = 8080\n")
        .expect("toml_to_data_bin");
    let out = data_bin_to_xml(&reframe(&payload)).expect("data_bin_to_xml");
    assert!(out.contains("localhost"));
    assert!(out.contains("8080"));
}
